extracters/cleaners.py

```diff
@@ -22,26 +22,50 @@ def share_parse(value):
 
 def ymd_to_datetime(year, month, day, which="begin"):
 
-	# These might raise ValueError if the string is not an int
-	# which is fine at this stage
+
 	if type(year) is not int:
-		year = int(year)
+		try:
+			year = int(year)
+		except:
+			print("DATE CLEAN: year is %r" % year)
+			year = None
+
 	if type(month) is not int:
-		month = int(month)
+		try:
+			month = int(month)
+		except:
+			print("DATE CLEAN: month is %r" % month)
+			month = None
+
 	if type(day) is not int:
-		day = int(day)
+		try:
+			day = int(day)
+		except:
+			day = None
+
+	if not year:
+		if which == "begin":
+			year = 0
+		else:
+			year = 9999
 
-	if not month:
+	if not month or month > 12 or month < 0:
 		if which == "begin":
 			month = 1
 		else:
 			month = 12
-	if not day:
+
+	try:
```

Comment:
`calendar.monthrange` seems to only raise an exception if `month` is not a valid value (not in the range 1–12). Given appropriate validation of `month` above, should we assume that `calendar.monthrange` will not `raise` instead of falling back to `31`? The fallback might be correct, but feels overly defensive.

Reply:
Yeah, I anticipate month being 0, or 13, or all sorts of other garbage in the data. For a cleaning function, I'd prefer it be as defensive as possible.

e.g.

```
>>> calendar.monthrange(1900, 0)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/calendar.py", line 120, in monthrange
    raise IllegalMonthError(month)
calendar.IllegalMonthError: bad month number 0; must be 1-12
```

Reply:
But shouldn't the code directly above this already do that validation? `calendar.monthrange(1900, 0)` could never be called here, because `month=0` can never pass the previous code, right?

Reply:
I'm ok with this as is, but want to make sure I'm understanding both the intention and what the code is actually doing.

Reply:
Yes, you're right :) Removing the try/except